addons/website_event/models/event_event.py

```diff
@@ -116,28 +141,66 @@ def _compute_is_participating(self):
+        events_in_self = [('event_id', 'in', self.ids)] if self else []
         if self.env.user._is_public() and not current_visitor:
-            events = self.env['event.event']
-        elif self.env.user._is_public():
-            events = self.env['event.registration'].sudo().search(
-                expression.AND([base_domain, [('visitor_id', '=', current_visitor.id)]])
-            ).event_id
+            return self.env['event.event']
+        if self.env.user._is_public():
+            domain = expression.AND([[
+                ('state', '!=', 'cancel'),
+                ('visitor_id', '=', current_visitor.id),
+            ], events_in_self])
         else:
             if current_visitor:
-                domain = [
+                partner_domain = [
                     '|',
                     ('partner_id', '=', self.env.user.partner_id.id),
                     ('visitor_id', '=', current_visitor.id)
                 ]
             else:
-                domain = [('partner_id', '=', self.env.user.partner_id.id)]
-            events = self.env['event.registration'].sudo().search(
-                expression.AND([base_domain, domain])
-            ).event_id
+                partner_domain = [('partner_id', '=', self.env.user.partner_id.id)]
+            domain = expression.AND([
+                partner_domain,
+                expression.AND([[('state', '!=', 'cancel')], events_in_self])
+            ])
+
+        return self.env['event.registration'].sudo().search(domain).event_id
+
+    @api.depends_context('uid')
+    @api.depends('visibility')
+    def _compute_can_see_on_website(self):
+        user = self.env.user
+        events_user_participating = self._get_events_where_user_participates()
+        for record in self:
+            if record.visibility == 'public' or record in events_user_participating:
+                record.can_see_on_website = True
+            elif not user._is_public() and record.visibility == 'logged_users':
+                record.can_see_on_website = True
+            else:
+                record.can_see_on_website = False
+
+    @api.model
+    def _search_can_see_on_website(self, operator, value):
+        if operator not in ['=', '!=']:
+            raise NotImplementedError(_('This operator is not supported'))
+        if not isinstance(value, bool):
+            raise UserError(_('Value should be True or False (not %)', value))
+        check_has_access = operator == '=' and value or operator == '!=' and not value
+        user = self.env.user
+        domain = [('is_participating', '=', True)]
+
+        if not user._is_public():
+            domain = expression.OR([domain, [('visibility', 'in', ['public', 'logged_users'])]])
+        else:
+            domain = expression.OR([domain, [('visibility', '=', 'public')]])
+
+        events = self.env['event.event'].search(domain)
+        return [('id', 'in' if check_has_access else 'not in', events.ids)]
```

Comment:
Can we use an "_search" to leverage the ORM rather than fetching all the events?

Could also be a good candidate for the new "any" and "not any" operators maybe.